### PR TITLE
`superagent` replacement and web worker for web

### PR DESCRIPTION
**Pull request by parfeon**
feat(network): replace `superagent` with `fetch`

Upgraded the network layer, replacing the `superagent` module with the `Fetch API` for browser integrations and node-fetch for `npm` integrations, ensuring enhanced performance and reliability.

feat(service-worker): added service worker 

Enhanced the subscribe feature with service worker support, improving user experience across multiple browser windows and tabs.
    
feat(types): client interface rewritten with TypeScript
    
The client interface rewritten with TypeScript, which gives an up-to-date interface.

### CODE REVIEW

File: src/web/configuration.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * If the browser fails to detect the network changes from Wi-Fi to LAN and vice versa, or you get
+ * reconnection issues, set the flag to `false`. This allows the SDK reconnection logic to take over.
+ */
+const LISTEN_TO_BROWSER_NETWORK_EVENTS = true;
+
+/**
+ * Whether PubNub client should try utilize existing TCP connection for new requests or not.
+ */
+const KEEP_ALIVE = true;
```
Comment (mohitpubnub) [inline src/web/configuration.ts:23]: Great! to keep it true default.
also, assuming that it wont have any side effect.

File: src/web/index.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ */
+export default class PubNub extends PubNubCore<ArrayBuffer | string, PubNubFileParameters, PubNubFile> {
+  /**
+   * Data encryption / decryption module constructor.
+   */
+  static CryptoModule = WebCryptoModule;
+
+  constructor(configuration: PubNubConfiguration) {
+    const configurationCopy = setDefaults(configuration);
+    const platformConfiguration = { ...configurationCopy, sdkFamily: 'Nodejs', PubNubFile };
```
Comment (mohitpubnub) [inline src/web/index.ts:32]: Shouldn’t it be ‘web’ instead of `‘Nodejs’`?

> sdkFamily: 'Web'

Comment (mohitpubnub) [inline src/web/index.ts:32]:
```suggestion
    const platformConfiguration = { ...configurationCopy, sdkFamily: 'Web', PubNubFile };
```

File: src/web/configuration.ts (hunk 1 above)
Comment (parfeon) [reply src/web/configuration.ts:23]: It looks like `true` is the default for `fetch` Web API itself in web environment.

File: src/web/index.ts (hunk 2 above)
Comment (parfeon) [reply src/web/index.ts:32]: Oh shoot, probably didn't save when fixed it once. Thank you for noticing!

File: src/transport/web-transport.ts [hunk 3]
```diff
@@ -77,13 +77,17 @@ export class WebTransport implements Transport {
...
-      },
-    );
+    // this.worker = new Worker(
+    //   URL.createObjectURL(new Blob([WEB_WORKER_PLACEHOLDER], { type: 'application/javascript' })),
+    //   {
+    //     name: '/pubnub',
+    //     type: 'module',
+    //   },
+    // );
+    this.worker = new Worker(WEB_WORKER_PLACEHOLDER, {
```
Comment (mohitpubnub) [inline src/transport/web-transport.ts:80]: 👍🏻

**Review by mohitpubnub — APPROVED** [review]
So far Look good!

File: lib/core/components/subscription-manager.js [hunk 4]
```diff
@@ -35,8 +29,8 @@ export class SubscriptionManager {
         this.presenceChannels = {};
         this.heartbeatTimer = null;
         this.presenceState = {};
-        this.pendingChannelGroupSubscriptions = [];
-        this.pendingChannelSubscriptions = [];
+        this.pendingChannelGroupSubscriptions = new Set();
```
Comment (mohitpubnub) [inline lib/core/components/subscription-manager.js:32]: 👍🏻 . This was pointed out some time back but missed!!
Good catch!!

File: tsconfig.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "$schema": "https://json.schemastore.org/tsconfig",
   "compilerOptions": {
     "target": "ES6",
-    "module": "ES6",
+    "module": "CommonJS",
```
Comment (mohitpubnub) [inline tsconfig.json:5]: 👍🏻

**Review by mohitpubnub — APPROVED** [review]

**Comment by k-funk** [comment]
this change suggests that the minimum browser versions supported may have been bumped to 2017. https://caniuse.com/fetch

Does pubnub document anywhere what these minimums are?

**Comment by parfeon** [comment]
@k-funk The latest one, Safari, jumped on the train of modern techs, did it seven years ago, while others did it nine years ago. To keep compatibility with the most number of user environments, we deliberately didn't use the most advanced browser features.
We don't have a separate documentation section which specifies the minimum browser version, but because it has been adopted almost a decade ago, we didn't felt that we need to add it. If users were using these outdated browsers or operating systems, they would have issues with modern web applications anyway.

**Pull request merged**